sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -206,13 +206,15 @@ object JoinReorderDP extends PredicateHelper with Logging {
       filters: Option[JoinGraphInfo]): JoinPlanMap = {
 
     val nextLevel = new JoinPlanMap
-    var k = 0
     val lev = existingLevels.length - 1
+    var k = lev
     // Build plans for the next level from plans at level k (one side of the join) and level
     // lev - k (the other side of the join).
-    // For the lower level k, we only need to search from 0 to lev - k, because when building
+    // For the higher level k, we only need to search from lev to lev - k, because when building
     // a join from A and B, both A J B and B J A are handled.
-    while (k <= lev - k) {
+    // Start searching from highest level to make sure that optimally ordered input doesn't get
+    // reordered into another plan with the same cost.
```

Comment:
Could you describe more about why we need to change this search order for making this rule idempotent? If how to update next-level join plans is a root cause, we cannot simply modify the update logic below when different plans having the same cost found?

https://github.com/apache/spark/blob/d15f504a5e8bd8acfb6dc1ee138f7d92ff211396/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L227-L233

Reply:
This was also my first idea, but I couldn't figure it out.

I'll try to give two examples, where all pairwise joins have the same cost and we can join only plans with consecutive letters (AB, but bot AC)
Keep in mind, that we try build left-deep trees:
https://github.com/apache/spark/blob/4619acc3ce72a6df9ac849d045ecdbed56a562be/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L301

Let the inital join be ((AB)C)

Firstly if we kept the first canditate 

1. iteration (the initial state):
(A)  (B)  (C)   

2. iteration (nested loop over the single elements):
(AB)  (BC)   
Disgarded (BA) and (CB), because they were late

3. iteration (outer loop is single element, inner is two element):
((BC)A)
Disgarded ((AB)C) because it was late

Secondly if we kept the last one
1. iteration (the initial state):
(A)  (B)  (C)   

2. iteration (nested loop over the single elements):
(BA)  (CB)   
Disgarded (AB) and (BC), because they were overwritten

3. iteration (outer loop is single element, inner is two element):
((BA)C)
Disgarded ((CB)A), because it was overwritten

Neither of these gave the original result. But lets try iterating from the largest (while keeping the first candidate)

1. iteration (the initial state):
(A)  (B)  (C)   
The initial state

2. iteration (nested loop over the single elements):
(AB)  (BC)   
Disgarded (BA) and (CB), because they were late

3. iteration (outer loop is two element, inner is single element):
((AB)C)
Disgarded ((BC)A) because it was late

Now, if we would add another plan D, then in the 4. iteration it would be appended to the ((AB)C) and we would still have correct result.

Note that if the input would be ((AB)(CD)), then the result would also be (((AB)C)D), but we must be stable only on second iteration and because we output left-deep trees, then we must be stable on left-deep trees, to be idempotent. 

Also by left-deep we mean "as left-deep" as possible. 

And finally, yes, I know that this dummy example is no rigorous proof, but we have 2 things going for us:
1. Out of all the test cases we have, none broke this rule
2. The idempotentsi test is used only in UTs, so there is no chance of breaking the user side by marking this as  idempotent.

Reply:
I pondered about this a bit more. I think this is a bit more general explanaiton:

After the first round of optimisation the plans are ordered: ABCD....
Lets ignore the brackets to keep it general.
On the second run we just have to redraw the brackets without changing the order.

We know that the result is as left-deep as possible. 

So, if we would be iterating from the smaller ones, then we could have (K) and (LM), that we want to join, but because it has to be left-deep, we must order it as ((LM)K).

But, if we start from the bigger ones, then we hit (KL) and (M) first and joining these will not reorder the letters.

For stability I think, that there are at least 4 factors, that must match with each others:
How do we join: left-deep, right-deep....
The iteration order in one "level": left-to-right, right-to-left.
The order of picking levels: bigger first, smaller first.
Which one we keep, when joins have equal cost: first, last...

There might be more. Changing the order of picking levels gets us to stability in one change.
There are most definitely more ways to achieve stability, but it seems, that it would require more than one change.

Reply:
Thanks for your detailed explanation. IMO a fix should be simple and intuitive in terms of maintainability. Another idea that I came up with after I read the description above was to simply sort input candidate plans by some values (e.g., `semanticHash`) at the beginning of `search`. It seems there are multiple options to solve this (like the four factors you described above). Any reason to choose the current approach out from them? You chose the simplest one?

Reply:
Btw, could you update the PR description based on the explanation above? That explanation looks useful for making others understood smoothly, I think.

Reply:
One more comment; could you add some tests based on the explanation scenario above?

Reply:
To my surprise the `semanticHash` changes between executions, so sorting by it makes the results of the `PlanStabilitySuite` not stable. 
In the second commit of this PR I tried the sorting, but could not get it to work: https://github.com/apache/spark/pull/29871/commits/d58966ce5434d50044e8d610978f075cbcad92b1

Reply:
> To my surprise the semanticHash changes between executions, so sorting by it makes the results of the PlanStabilitySuite not stable.

Ah, I see... This rule rewrites candidate plans after reordering plans, so IIUC `semanticHash` values might change in the second run...

> TBH It wasn't my first idea, but it is the first one, I got to work - the other ones seemed to require too much code change and I did not go further with those.

okay, let's wait for the comments of other reviewers.

Reply:
> Ah, I see... This rule rewrites candidate plans after reordering plans, so IIUC `semanticHash` values might change in the second run...
> 

Yeah, that was one issue, that complicated the approach.

But, what made it even worse is that the value for `semanticHash` will change every time you restart the JVM. I'm not sure if its intended or not and I could not find the source of randomness.

Reply:
> But, what made it even worse is that the value for semanticHash will change every time you restart the JVM. I'm not sure if its intended or not and I could not find the source of randomness.

Yea, that's a design and why do we need to keep the same `semanticHash` values between different JVMs?

Reply:
> Yea, that's a design and why do we need to keep the same `semanticHash` values between different JVMs?

It does not allow us sort by it and then compare results in the `PlanStbilitySuite`.

Reply:
Ah, I see. I only cared about the idempotence check in `RuleExecutor`.